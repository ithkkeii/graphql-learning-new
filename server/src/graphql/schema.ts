const { gql } = require("apollo-server-express");

export const typeDefs = gql`
  # interface MutationResponse {
  #   code: String!
  #   success: Boolean!
  #   message: String!
  # }

  type AuthPayload {
    token: String!
    user: User!
  }

  type TestCase {
    text: String!
    testString: String!
  }

  type Challenge {
    id: ID!
    title: String!
    content: String!
    level: String!
    points: Int!
    contest: Contest!
    testCases: [TestCase!]!
    testInputs: [String!]!
    challengeSeed: String!
    passedUser: [User!]!
  }

  type User {
    id: ID!
    username: String!
    password: String
    firstname: String!
    lastname: String!
    solvedChallenges: [Challenge!]!
    likedChallenges: [Challenge!]!
    totalPoints: Int!
  }

  type Contest {
    id: ID!
    name: String!
    challenges: [Challenge!]!
  }

  type Query {
    getChallenge(id: ID!): Challenge
    getChallenges: [Challenge!]!
    getUser(id: ID!): User
    getUsers: [User!]!
    getContest(id: ID!): Contest!
    getContests: [Contest!]!
  }

  input UserInput {
    username: String!
    password: String!
    firstname: String!
    lastname: String!
  }

  input TestCaseInput {
    text: String!
    testString: String!
  }

  input ChallengeInput {
    title: String!
    content: String!
    level: String!
    points: Int!
    contestId: ID!
    testCases: [TestCaseInput!]!
    testInputs: [String!]!
    challengeSeed: String!
  }

  input ContestInput {
    name: String!
  }

  type Mutation {
    login(username: String!, password: String!): String!
    register(user: UserInput!): AuthPayload!
    editUser(userId: ID!, user: UserInput!): User!
    addChallenge(challenge: ChallengeInput!): Challenge!
    editChallenge(challengeId: ID!, challenge: ChallengeInput!): Challenge!
    addOrRemoveSolvedChallenges(userId: ID!, challengeId: ID!): [ID!]!
    addOrRemoveLikedChallenges(userId: ID!, challengeId: ID!): [ID!]!
    addContest(contest: ContestInput!): Contest!
    editContest(contestId: ID!, contest: ContestInput!): Contest!
    submitAnswer(challengeId: ID!, answer: String!): [String]!
  }
`;
