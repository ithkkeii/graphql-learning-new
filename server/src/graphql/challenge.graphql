type Challenge {
  id: ID!
  title: String!
  content: String!
  level: String!
  points: Int!
  category: Category!
  passedUser: [User!]!
}

type User {
  id: ID!
  username: String!
  password: String
  firstname: String!
  lastname: String!
  solvedChallenges: [Challenge!]!
  likedChallenges: [Challenge!]!
  totalPoints: Int!
}

type Category {
  id: ID!
  name: String!
  challenges: [Challenge!]!
}
