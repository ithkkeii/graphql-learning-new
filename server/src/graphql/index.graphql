# import Challenge from './challenge.graphql'
# import User from './challenge.graphql'
# import Category from './challenge.graphql'

type Query {
  getChallenge(id: ID!): Challenge
  getChallenges: [Challenge!]!
  getUser(id: ID!): User
  getUsers: [User!]!
  getCategory: Category!
  getCategories: [Category!]!
}

input UserInput {
  username: String!
  password: String!
  firstname: String!
  lastname: String!
}

input ChallengeInput {
  title: String!
  content: String!
  level: String!
  categoryId: ID!
}

input CategoryInput {
  name: String!
}

type Mutation {
  addUser(user: UserInput!): User!
  editUser(userId: ID!, user: UserInput!): User!
  addChallenge(challenge: ChallengeInput!): Challenge!
  editChallenge(challengeId: ID!, challenge: ChallengeInput!): Challenge!
  addOrRemoveSolvedChallenges(userId: ID!, challengeId: ID!): [ID!]!
  addOrRemoveLikedChallenges(userId: ID!, challengeId: ID!): [ID!]!
  addCategory(category: CategoryInput!): Category!
  editCategory(categoryId: ID!, category: CategoryInput!): Category!
}
