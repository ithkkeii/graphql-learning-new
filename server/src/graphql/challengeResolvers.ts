import { Challenge, validateChallenge } from "../models/challenge";
import { User } from "../models/user";
import { Contest } from "../models/contest";
import mongoose from "mongoose";
import { slugify } from "../utils/utilities";

const challengeResolvers = {
  Query: {
    getChallenge: async (parent: any, args: any, context: any) => {
      return await Challenge.findOne({ slug: args.slug });
    },
    getChallenges: async (parent: any, args: any, context: any) => {
      return await Challenge.find({});
    },
  },
  Challenge: {
    passedUser: async (parent: any, args: any, context: any, info: any) => {
      return await User.find({ solvedChallenges: parent.id }, "-password");
    },
    contest: async (parent: any, args: any, context: any, info: any) => {
      return await Contest.findById(parent.contest);
    },
  },
  Mutation: {
    addChallenge: async (parent: any, args: any, context: any, info: any) => {
      const {
        title,
        content,
        level,
        points,
        contestId,
        testCases,
        testInputs,
        challengeSeed,
      } = args.challenge;

      const { error } = validateChallenge(args.challenge);
      if (error) throw new Error(error.details[0].message);

      // If title exist
      let challenge = await Challenge.findOne({ title });
      if (challenge) throw new Error(`Title is already taken`);

      // Create a unique slug
      const slug = slugify(title);
      const checkSlug = await Challenge.findOne({ slug });
      if (checkSlug) throw new Error(`Title creates an existed slug`);

      const contest = await Contest.findById(contestId);
      if (!contest) throw new Error(`Invalid contest's id`);

      challenge = new Challenge({
        title,
        slug,
        content,
        level,
        points,
        contest: contestId,
        testCases,
        testInputs,
        challengeSeed,
      });
      challenge = await challenge.save();

      return challenge;
    },
    editChallenge: async (parent: any, args: any, context: any, info: any) => {
      const { challengeId } = args;
      const {
        title,
        content,
        level,
        points,
        contestId,
        testCases,
        testInputs,
        challengeSeed,
      } = args.challenge;

      const { error } = validateChallenge(args.challenge);
      if (error) throw new Error(error.details[0].message);

      if (!mongoose.Types.ObjectId.isValid(challengeId))
        throw new Error(`Invalid challenge's id`);

      let challenge = await Challenge.findById(challengeId);
      if (!challenge) throw new Error(`Invalid challenge's id`);

      const uniqueTitleChallenge = await Challenge.findOne({
        title,
        _id: {
          $nin: [challengeId],
        },
      });
      if (uniqueTitleChallenge)
        throw new Error(`Challenge's title is already exist`);

      // Create a unique slug
      const slug = slugify(title);
      const checkSlug = await Challenge.findOne({
        slug,
        _id: {
          $nin: [challengeId],
        },
      });
      if (checkSlug) throw new Error(`Title creates an existed slug`);

      const contest = await Contest.findById(contestId);
      if (!contest) throw new Error(`Invalid contest's id`);

      challenge.title = title;
      challenge.content = content;
      challenge.level = level;
      challenge.points = points;
      challenge.contest = contestId;
      challenge.testCases = testCases;
      challenge.testInputs = testInputs;
      challenge.challengeSeed = challengeSeed;

      challenge = await challenge.save();
      return challenge;
    },
    // ! Dont need to remove solvedChallenges
    addOrRemoveSolvedChallenges: async (
      parent: any,
      args: any,
      context: any
    ) => {
      const { userId, challengeId } = args;

      if (!mongoose.Types.ObjectId.isValid(challengeId))
        throw new Error(`Invalid challenge's id`);

      if (!mongoose.Types.ObjectId.isValid(userId))
        throw new Error(`Invalid user's id`);

      let user = await User.findById(userId);
      if (!user) throw new Error(`Invalid user's id`);

      let challenge = await Challenge.findById(challengeId);
      if (!challenge) throw new Error(`Invalid challenge's id`);

      const index: number = user.solvedChallenges.indexOf(challengeId);
      if (index === -1) {
        user.solvedChallenges.push(challengeId);
      } else {
        user.solvedChallenges.splice(index, 1);
      }
      user = await user.save();
      delete user.account.password;
      return user.solvedChallenges;
    },
    addOrRemoveLikedChallenges: async (
      parent: any,
      args: any,
      context: any
    ) => {
      const { userId, challengeId } = args;

      if (!mongoose.Types.ObjectId.isValid(challengeId))
        throw new Error(`Invalid challenge's id`);

      if (!mongoose.Types.ObjectId.isValid(userId))
        throw new Error(`Invalid user's id`);

      let user = await User.findById(userId);
      if (!user) throw new Error(`Invalid user's id`);

      let challenge = await Challenge.findById(challengeId);
      if (!challenge) throw new Error(`Invalid challenge's id`);

      const index: number = user.likedChallenges.indexOf(challengeId);
      if (index === -1) {
        user.likedChallenges.push(challengeId);
      } else {
        user.likedChallenges.splice(index, 1);
      }
      user = await user.save();
      delete user.account.password;
      return user.likedChallenges;
    },
  },
};

export default challengeResolvers;
