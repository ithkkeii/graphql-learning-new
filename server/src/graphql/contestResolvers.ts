import { Contest, validateContest } from "../models/contest";
import { Challenge } from "../models/challenge";
import mongoose from "mongoose";
import { createWriteStream } from "fs";
import { code } from "../code";

const contestResolvers = {
  Query: {
    getContest: async (parent: any, args: any, context: any) => {
      return await Contest.findById(args.id);
    },
    getContests: async (parent: any, args: any, context: any) => {
      return await Contest.find({});
    },
  },
  Contest: {
    challenges: async (parent: any, args: any, context: any, info: any) => {
      return await Challenge.find({ contest: parent._id });
    },
  },
  Mutation: {
    addContest: async (parent: any, args: any, context: any, info: any) => {
      const { name } = args.contest;

      const { error } = validateContest(args.contest);
      if (error) throw new Error(error.details[0].message);

      // Check name unique
      let contest = await Contest.findOne({ name });
      if (contest) throw new Error(`Name of contest is already taken`);

      contest = new Contest({ name });
      return await contest.save();
    },
    editContest: async (parent: any, args: any, context: any, info: any) => {
      const { name } = args.contest;

      const { error } = validateContest(args.contest);
      if (error) throw new Error(error.details[0].message);

      if (!mongoose.Types.ObjectId.isValid(args.contestId))
        throw new Error(`Invalid contest's id`);

      let contest = await Contest.findById(args.contestId);
      if (!contest) throw new Error(`Invalid contest's id`);

      // Check name unique
      let uniqueContest = await Contest.findOne({
        name,
        _id: { $nin: [args.contestId] },
      });
      if (uniqueContest) throw new Error(`Name of contest is already taken`);

      contest.name = name;
      return await contest.save();
    },
    submitAnswer: async (parent: any, args: any, context: any, info: any) => {
      const { challengeId, answer } = args;

      const sum: Function = new Function(answer)();
      const checkAnonymous: Function = new Function(
        "assert",
        "sum",
        "return assert(sum(1,2) === 3)"
      );

      // await writeFile("test.ts", code, () => {});
      const writer = createWriteStream("test.ts", { encoding: "utf8" });
      writer.write(code);
      // if (anonymousFunc()(1, 2) === 3) return [true];

      // console.log(checkAnonymous(assert, sum).toString());
      console.log(sum.toString());
      return [true];
    },
  },
};

export default contestResolvers;
