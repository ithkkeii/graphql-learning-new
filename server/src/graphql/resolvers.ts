import mongoose, { mongo } from "mongoose";
import { Challenge, validateChallenge } from "../models/challenge";
import { User, validateUser } from "../models/user";
import { Contest, validateContest } from "../models/contest";
import { sign } from "jsonwebtoken";
import { authenticateUser } from "../utils/auth";
import { assert } from "chai";
import { writeFile, createWriteStream, write } from "fs";
import { join } from "path";
import { code } from "../code";
const bcrypt = require("bcrypt");

const resolvers = {
  Query: {
    getChallenge: async (parent: any, args: any, context: any) => {
      return await Challenge.findById(args.id);
    },
    getChallenges: async (parent: any, args: any, context: any) => {
      return await Challenge.find({});
    },
    //! Need admin authorization
    getUser: async (parent: any, args: any, context: any) => {
      return await User.findById(args.id);
    },
    //! Need admin authorization
    getUsers: async (parent: any, args: any, context: any) => {
      const user = await authenticateUser(context);
      return await User.find({});
    },
    getContest: async (parent: any, args: any, context: any) => {
      return await Contest.findById(args.id);
    },
    getContests: async (parent: any, args: any, context: any) => {
      return await Contest.find({});
    },
  },
  User: {
    solvedChallenges: async (
      parent: any,
      args: any,
      context: any,
      info: any
    ) => {
      const getChallenge = async (challengeId: String) => {
        return await Challenge.findById(challengeId);
      };

      const getChallenges = async () => {
        return Promise.all(
          parent.solvedChallenges.map((challengeId: any) =>
            getChallenge(challengeId)
          )
        );
      };

      return await getChallenges();
    },
    likedChallenges: async (
      parent: any,
      args: any,
      context: any,
      info: any
    ) => {
      const getChallenge = async (challengeId: String) => {
        return await Challenge.findById(challengeId);
      };

      const getChallenges = async () => {
        return Promise.all(
          parent.likedChallenges.map((challengeId: any) =>
            getChallenge(challengeId)
          )
        );
      };

      return await getChallenges();
    },
  },
  Challenge: {
    passedUser: async (parent: any, args: any, context: any, info: any) => {
      return await User.find({ solvedChallenges: parent.id }, "-password");
    },
    contest: async (parent: any, args: any, context: any, info: any) => {
      return await Contest.findById(parent.contest);
    },
  },
  Contest: {
    challenges: async (parent: any, args: any, context: any, info: any) => {
      return await Challenge.find({ contest: parent.contest });
    },
  },
  Mutation: {
    login: async (parent: any, args: any, context: any, info: any) => {
      const { username, password } = args;

      const { error } = validateUser({
        username,
        password,
        firstname: "validFirstName",
        lastname: "validLastName",
      });
      if (error) throw new Error(error.details[0].message);

      const user = await User.findOne({ username });
      if (!user) throw new Error(`Invalid username or password`);

      const isValidPassword = await bcrypt.compare(password, user.password);
      if (!isValidPassword) throw new Error(`Invalid username or password`);

      //! Import config for privatekey later & generateAuthToken need in user schema
      const token: string = sign(
        {
          _id: user._id,
          firstname: user.firstname,
          lastname: user.lastname,
        },
        "helloworld"
      );
      return token;
    },
    register: async (parent: any, args: any, context: any, info: any) => {
      const { username, password, firstname, lastname } = args.user;

      const { error } = validateUser(args.user);
      if (error) throw new Error(error.details[0].message);

      // If username already exist
      let user = await User.findOne({ username });
      if (user) throw new Error(`Username already taken`);

      user = new User({
        username,
        password,
        firstname,
        lastname,
      });

      user.password = await bcrypt.hash(user.password, 10);

      //? Do i need include password into token ?
      //! Import config for privatekey later & generateAuthToken need in user schema
      const token: string = sign(
        {
          _id: user._id,
          firstname: user.firstname,
          lastname: user.lastname,
        },
        "helloworld"
      );
      user = await user.save();
      delete user.password;

      return { token, user };
    },
    // ? How to create flexible InputUser graphql
    editUser: async (parent: any, args: any, context: any, info: any) => {
      const { userId } = args;
      const { username = "", password = "", firstname, lastname } = args.user;

      const { error } = validateUser(args.user);
      if (error) throw new Error(error.details[0].message);

      if (!mongoose.Types.ObjectId.isValid(userId))
        throw new Error(`Invalid user's id`);

      let user = await User.findById(userId);
      if (!user) throw new Error(`Invalid user's id`);

      user.firstname = firstname;
      user.lastname = lastname;
      user = await user.save();
      delete user.password;

      return user.save();
    },
    addChallenge: async (parent: any, args: any, context: any, info: any) => {
      const {
        title,
        content,
        level,
        points,
        contestId,
        testCases,
        testInputs,
        challengeSeed,
      } = args.challenge;

      const { error } = validateChallenge(args.challenge);
      if (error) throw new Error(error.details[0].message);

      const contest = await Contest.findById(contestId);
      if (!contest) throw new Error(`Invalid contest's id`);

      // If title exist
      let challenge = await Challenge.findOne({ title });
      if (challenge) throw new Error(`Title is already taken`);

      challenge = new Challenge({
        title,
        content,
        level,
        points,
        contest: contestId,
        testCases,
        testInputs,
        challengeSeed,
      });
      challenge = await challenge.save();

      return challenge;
    },
    editChallenge: async (parent: any, args: any, context: any, info: any) => {
      const { challengeId } = args;
      const {
        title,
        content,
        level,
        points,
        contestId,
        testCases,
        testInputs,
        challengeSeed,
      } = args.challenge;

      const { error } = validateChallenge(args.challenge);
      if (error) throw new Error(error.details[0].message);

      if (!mongoose.Types.ObjectId.isValid(challengeId))
        throw new Error(`Invalid challenge's id`);

      let challenge = await Challenge.findById(challengeId);
      if (!challenge) throw new Error(`Invalid challenge's id`);

      const uniqueTitleChallenge = await Challenge.findOne({
        title,
        _id: { $nin: [challengeId] },
      });
      if (uniqueTitleChallenge)
        throw new Error(`Challenge' title is already exist`);

      const contest = await Contest.findById(contestId);
      if (!contest) throw new Error(`Invalid contest's id`);

      challenge.title = title;
      challenge.content = content;
      challenge.level = level;
      challenge.points = points;
      challenge.contest = contestId;
      challenge.testCases = testCases;
      challenge.testInputs = testInputs;
      challenge.challengeSeed = challengeSeed;

      challenge = await challenge.save();
      return challenge;
    },
    // ! Dont need to remove solvedChallenges
    addOrRemoveSolvedChallenges: async (
      parent: any,
      args: any,
      context: any
    ) => {
      const { userId, challengeId } = args;

      if (!mongoose.Types.ObjectId.isValid(challengeId))
        throw new Error(`Invalid challenge's id`);

      if (!mongoose.Types.ObjectId.isValid(userId))
        throw new Error(`Invalid user's id`);

      let user = await User.findById(userId);
      if (!user) throw new Error(`Invalid user's id`);

      let challenge = await Challenge.findById(challengeId);
      if (!challenge) throw new Error(`Invalid challenge's id`);

      const index: number = user.solvedChallenges.indexOf(challengeId);
      if (index === -1) {
        user.solvedChallenges.push(challengeId);
      } else {
        user.solvedChallenges.splice(index, 1);
      }
      user = await user.save();
      delete user.password;
      return user.solvedChallenges;
    },
    addOrRemoveLikedChallenges: async (
      parent: any,
      args: any,
      context: any
    ) => {
      const { userId, challengeId } = args;

      if (!mongoose.Types.ObjectId.isValid(challengeId))
        throw new Error(`Invalid challenge's id`);

      if (!mongoose.Types.ObjectId.isValid(userId))
        throw new Error(`Invalid user's id`);

      let user = await User.findById(userId);
      if (!user) throw new Error(`Invalid user's id`);

      let challenge = await Challenge.findById(challengeId);
      if (!challenge) throw new Error(`Invalid challenge's id`);

      const index: number = user.likedChallenges.indexOf(challengeId);
      if (index === -1) {
        user.likedChallenges.push(challengeId);
      } else {
        user.likedChallenges.splice(index, 1);
      }
      user = await user.save();
      delete user.password;
      return user.likedChallenges;
    },
    addContest: async (parent: any, args: any, context: any, info: any) => {
      const { name } = args.contest;

      const { error } = validateContest(args.contest);
      if (error) throw new Error(error.details[0].message);

      // Check name unique
      let contest = await Contest.findOne({ name });
      if (contest) throw new Error(`Name of contest is already taken`);

      contest = new Contest({ name });
      return await contest.save();
    },
    editContest: async (parent: any, args: any, context: any, info: any) => {
      const { name } = args.contest;

      const { error } = validateContest(args.contest);
      if (error) throw new Error(error.details[0].message);

      if (!mongoose.Types.ObjectId.isValid(args.contestId))
        throw new Error(`Invalid contest's id`);

      let contest = await Contest.findById(args.contestId);
      if (!contest) throw new Error(`Invalid contest's id`);

      // Check name unique
      let uniqueContest = await Contest.findOne({
        name,
        _id: { $nin: [args.contestId] },
      });
      if (uniqueContest) throw new Error(`Name of contest is already taken`);

      contest.name = name;
      return await contest.save();
    },
    submitAnswer: async (parent: any, args: any, context: any, info: any) => {
      const { challengeId, answer } = args;

      const sum: Function = new Function(answer)();
      const checkAnonymous: Function = new Function(
        "assert",
        "sum",
        "return assert(sum(1,2) === 3)"
      );

      // await writeFile("test.ts", code, () => {});
      const writer = createWriteStream("test.ts", { encoding: "utf8" });
      writer.write(code);
      // if (anonymousFunc()(1, 2) === 3) return [true];

      // console.log(checkAnonymous(assert, sum).toString());
      console.log(sum.toString());
      return [true];
    },
  },
};

export default resolvers;
