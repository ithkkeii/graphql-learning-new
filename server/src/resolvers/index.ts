import userResolvers from "./userResolvers";
import challengeResolvers from "./challengeResolvers";
import contestResolvers from "./contestResolvers";

const resolvers = {
  Query: {
    ...challengeResolvers.Query,
    ...userResolvers.Query,
    ...contestResolvers.Query,
  },
  User: userResolvers.User,
  Challenge: challengeResolvers.Challenge,
  Contest: contestResolvers.Contest,
  Mutation: {
    ...userResolvers.Mutation,
    ...challengeResolvers.Mutation,
    ...contestResolvers.Mutation,
  },
};

export default resolvers;
