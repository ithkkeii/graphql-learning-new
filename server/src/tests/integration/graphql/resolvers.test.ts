import { gql, ApolloServer } from "apollo-server-express";
const { createTestClient } = require("apollo-server-testing");
import { typeDefs } from "../../../graphql/schema";
import resolvers from "../../../graphql/resolvers";
import { connect } from "mongoose";

connect("mongodb://localhost:27017/hacker-challenge", {
  useNewUrlParser: true,
  useUnifiedTopology: true,
})
  .then(() => console.log("Connected to mongoDB..."))
  .catch((err: { message: any }) =>
    console.log(`Cannot connected to mongoDB with err : ${err.message}`)
  );

const addUser = gql`
  input UserInput {
    username: String!
    password: String!
    firstname: String!
    lastname: String!
  }

  mutation AddUserMutation($user: UserInput!) {
    addUser(user: $user) {
      id
      username
      password
      firstname
      lastname
    }
  }
`;

const GET_USERS = gql`
  query GetUsers {
    getUsers {
      id
      username
      password
      firstname
      lastname
      totalPoints
      solvedChallenges {
        id
        title
        content
        level
        category {
          name
        }
        passedUser {
          id
          username
          password
          firstname
          lastname
          totalPoints
        }
      }
      likedChallenges {
        id
        title
        content
        level
        category {
          name
        }
        passedUser {
          id
          username
          password
          firstname
          lastname
          totalPoints
        }
      }
    }
  }
`;

const GET_USER = gql`
  query GetUser($id: ID!) {
    getUser(id: $id) {
      id
      username
      password
      firstname
      lastname
      totalPoints
      solvedChallenges {
        id
        title
        content
        level
        category {
          name
        }
        passedUser {
          id
          username
          password
          firstname
          lastname
          totalPoints
        }
      }
      likedChallenges {
        id
        title
        content
        level
        category {
          name
        }
        passedUser {
          id
          username
          password
          firstname
          lastname
          totalPoints
        }
      }
    }
  }
`;

const GET_CHALLENGE = gql`
  query GetChallenge($id: ID!) {
    getChallenge(id: $id) {
      id
      title
      content
      level
      points
      category {
        name
      }
      passedUser {
        id
        username
        password
        firstname
        lastname
        likedChallenges {
          id
        }
        solvedChallenges {
          id
        }
      }
    }
  }
`;

const GET_CHALLENGES = gql`
  query GetChallenges {
    getChallenges {
      id
      title
      content
      level
      points
      category {
        name
      }
      passedUser {
        id
        username
        password
        firstname
        lastname
        likedChallenges {
          id
        }
        solvedChallenges {
          id
        }
      }
    }
  }
`;

const GET_CATEGORIES = gql`
  query GetCategories {
    getCategories {
      name
      challenges {
        id
      }
    }
  }
`;

const GET_CATEGORY = gql`
  query GetCategory($id: ID!) {
    getCategory(id: $id) {
      name
      challenges {
        id
      }
    }
  }
`;

describe("Queries", () => {
  it("should return all users", async () => {
    const apolloServer = new ApolloServer({ typeDefs, resolvers });

    const { query, mutate } = createTestClient(apolloServer);

    const res = await query({ query: GET_USERS });
    expect(res).toMatchSnapshot();
  });

  it("should return single user with provided ID", async () => {
    const apolloServer = new ApolloServer({ typeDefs, resolvers });

    const { query, mutate } = createTestClient(apolloServer);

    const res = await query({
      query: GET_USER,
      variables: { id: "5e9beb8203a9480f22dcb733" },
    });
    expect(res).toMatchSnapshot();
  });

  it("should return all challenges", async () => {
    const apolloServer = new ApolloServer({ typeDefs, resolvers });

    const { query, mutate } = createTestClient(apolloServer);

    const res = await query({ query: GET_CHALLENGES });
    expect(res).toMatchSnapshot();
  });

  it("should return single challenge with provided ID", async () => {
    const apolloServer = new ApolloServer({ typeDefs, resolvers });

    const { query, mutate } = createTestClient(apolloServer);

    const res = await query({
      query: GET_CHALLENGE,
      variables: { id: "5e9bec5203a9480f22dcb737" },
    });
    expect(res).toMatchSnapshot();
  });

  it("should return all categories", async () => {
    const apolloServer = new ApolloServer({ typeDefs, resolvers });

    const { query, mutate } = createTestClient(apolloServer);

    const res = await query({
      query: GET_CATEGORIES,
      variables: { id: "5e9b2feccf07096f9ef2bc65" },
    });
    expect(res).toMatchSnapshot();
  });
});
