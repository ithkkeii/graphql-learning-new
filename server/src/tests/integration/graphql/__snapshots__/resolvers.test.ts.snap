// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries should return all categories 1`] = `
Object {
  "data": Object {
    "getCategories": Array [
      Object {
        "challenges": Array [],
        "name": "Play with maximun",
      },
      Object {
        "challenges": Array [],
        "name": "Play with tree",
      },
      Object {
        "challenges": Array [],
        "name": "Play with sort",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries should return all challenges 1`] = `
Object {
  "data": Object {
    "getChallenges": Array [
      Object {
        "category": Object {
          "name": "Play with tree",
        },
        "content": "xyzabc",
        "id": "5e9bec5203a9480f22dcb737",
        "level": "medium",
        "passedUser": Array [],
        "points": 0,
        "title": "Binary tree",
      },
      Object {
        "category": Object {
          "name": "Play with tree",
        },
        "content": "xyzabc",
        "id": "5e9bec6c03a9480f22dcb73e",
        "level": "medium",
        "passedUser": Array [
          Object {
            "firstname": "Voldemort",
            "id": "5e9beb8203a9480f22dcb733",
            "lastname": "Riddle",
            "likedChallenges": Array [
              Object {
                "id": "5e9bec6c03a9480f22dcb73e",
              },
            ],
            "password": null,
            "solvedChallenges": Array [
              Object {
                "id": "5e9bec6c03a9480f22dcb73e",
              },
            ],
            "username": "Voldemort",
          },
        ],
        "points": 0,
        "title": "Binary sorts",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries should return all users 1`] = `
Object {
  "data": Object {
    "getUsers": Array [
      Object {
        "firstname": "Voldemort",
        "id": "5e9beb8203a9480f22dcb733",
        "lastname": "Riddle",
        "likedChallenges": Array [
          Object {
            "category": Object {
              "name": "Play with tree",
            },
            "content": "xyzabc",
            "id": "5e9bec6c03a9480f22dcb73e",
            "level": "medium",
            "passedUser": Array [
              Object {
                "firstname": "Voldemort",
                "id": "5e9beb8203a9480f22dcb733",
                "lastname": "Riddle",
                "password": null,
                "totalPoints": 0,
                "username": "Voldemort",
              },
            ],
            "title": "Binary sorts",
          },
        ],
        "password": null,
        "solvedChallenges": Array [
          Object {
            "category": Object {
              "name": "Play with tree",
            },
            "content": "xyzabc",
            "id": "5e9bec6c03a9480f22dcb73e",
            "level": "medium",
            "passedUser": Array [
              Object {
                "firstname": "Voldemort",
                "id": "5e9beb8203a9480f22dcb733",
                "lastname": "Riddle",
                "password": null,
                "totalPoints": 0,
                "username": "Voldemort",
              },
            ],
            "title": "Binary sorts",
          },
        ],
        "totalPoints": 0,
        "username": "Voldemort",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries should return single challenge with provided ID 1`] = `
Object {
  "data": Object {
    "getChallenge": Object {
      "category": Object {
        "name": "Play with tree",
      },
      "content": "xyzabc",
      "id": "5e9bec5203a9480f22dcb737",
      "level": "medium",
      "passedUser": Array [],
      "points": 0,
      "title": "Binary tree",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries should return single user with provided ID 1`] = `
Object {
  "data": Object {
    "getUser": Object {
      "firstname": "Voldemort",
      "id": "5e9beb8203a9480f22dcb733",
      "lastname": "Riddle",
      "likedChallenges": Array [
        Object {
          "category": Object {
            "name": "Play with tree",
          },
          "content": "xyzabc",
          "id": "5e9bec6c03a9480f22dcb73e",
          "level": "medium",
          "passedUser": Array [
            Object {
              "firstname": "Voldemort",
              "id": "5e9beb8203a9480f22dcb733",
              "lastname": "Riddle",
              "password": null,
              "totalPoints": 0,
              "username": "Voldemort",
            },
          ],
          "title": "Binary sorts",
        },
      ],
      "password": null,
      "solvedChallenges": Array [
        Object {
          "category": Object {
            "name": "Play with tree",
          },
          "content": "xyzabc",
          "id": "5e9bec6c03a9480f22dcb73e",
          "level": "medium",
          "passedUser": Array [
            Object {
              "firstname": "Voldemort",
              "id": "5e9beb8203a9480f22dcb733",
              "lastname": "Riddle",
              "password": null,
              "totalPoints": 0,
              "username": "Voldemort",
            },
          ],
          "title": "Binary sorts",
        },
      ],
      "totalPoints": 0,
      "username": "Voldemort",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
